server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/AllocationActionListenerTests.java

```diff
@@ -59,4 +67,84 @@ public void testShouldDelegateFailureImmediately() {
 
         assertThat(completed.get(), equalTo(true));
     }
+
+    public void testShouldExecuteWithCorrectContext() {
+
+        var context = new ThreadContext(Settings.EMPTY);
+
+        // should not be changed after the listener is created
+        context.putHeader("header", "root");// ensure this is visible in the listener
+        context.addResponseHeader("header", "1");
+
+        var listener = new AllocationActionListener<>(ActionListener.wrap(ignore -> {
+            assertThat(context.getResponseHeaders().get("header"), containsInAnyOrder("1", "3", "4"));
+            assertThat(context.getHeader("header"), equalTo("root"));
```

Comment:
I'd prefer to have some verification that this listener is actually completed (even though a bug of that nature would make other tests fail too)